### PR TITLE
Java: IPA the CFG (second try)

### PR DESCRIPTION
**Pull request by owen-mc**
This converts the control flow graph to an IPA type to give us some more options in how we model control flow. There should be no change in any query results.

This resurrects [this PR](https://github.com/github/codeql/pull/711) from almost six years ago. The CFG code hasn't changed much in the interim, thankfully.

The commit history is pretty garbled, so it's probably best not to review commit-by-commit.

### CODE REVIEW

File: java/ql/lib/semmle/code/java/ControlFlowGraph.qll [hunk 1]
```diff
@@ -84,45 +84,113 @@ private import Completion
...
 
-/** A node in the expression-level control-flow graph. */
-class ControlFlowNode extends Top, @exprparent {
-  /** Gets the statement containing this node, if any. */
-  Stmt getEnclosingStmt() {
-    result = this or
-    result = this.(Expr).getEnclosingStmt()
+module ControlFlow {
+  private predicate hasControlFlow(Expr e) {
+    not e.getEnclosingStmt() instanceof ConstCase and
```
Comment (yoff) [inline java/ql/lib/semmle/code/java/ControlFlowGraph.qll:89]: This is interesting..

Comment (owen-mc) [reply java/ql/lib/semmle/code/java/ControlFlowGraph.qll]: In a line like `case "a" -> null;` I think this is trying to exclude the `"a"` but is accidentally excluding the `null` as well. I guess this wasn't an issue until switch expressions got added.

Comment (owen-mc) [reply java/ql/lib/semmle/code/java/ControlFlowGraph.qll]: Or maybe it's the arrow syntax that caused the problem, as extra expressions are now part of this statement.

**Comment by owen-mc** [comment]
The language test failure can only be fixed with an internal PR to update a .expected file. I will prepare one when this PR has been approved.

**Comment by owen-mc** [comment]
DCA has finished and it shows a mild analysis time increase. I ran some queries on apache/flink using before and after commits and looked at the tuple counts and predicate timings and didn't see anything obviously taking much longer. I don't think it's a bad join order or anything like that. But I guess that adding another layer will make analysis take a little longer.

File: java/ql/lib/semmle/code/java/ControlFlowGraph.qll [hunk 2]
```diff
@@ -4,7 +4,7 @@
  * The only API exported by this library are the toplevel classes `ControlFlowNode`
```
Comment (aschackmull) [inline java/ql/lib/semmle/code/java/ControlFlowGraph.qll:4]:
```suggestion
 * The only API exported by this library are the toplevel classes `ControlFlow::Node`
```

File: java/ql/lib/semmle/code/java/ControlFlowGraph.qll [hunk 3]
```diff
@@ -84,45 +84,122 @@ private import Completion
...
+    ExprParent getAstNode() {
+      result = this.asExpr() or
+      result = this.asStmt() or
+      this = TExitNode(result)
+    }
   }
 
-  /** Gets the basic block that contains this node. */
-  BasicBlock getBasicBlock() { result.getANode() = this }
+  /** A synthetic node for the exit of a callable. */
```
Comment (aschackmull) [inline java/ql/lib/semmle/code/java/ControlFlowGraph.qll:179]:
```suggestion
  /** A control flow node indicating the termination of a callable. */
```

File: java/ql/lib/semmle/code/java/ControlFlowGraph.qll [hunk 4]
```diff
@@ -853,15 +925,18 @@ private module ControlFlowGraphImpl {
...
     exists(InstanceOfExpr ioe | ioe.isPattern() and ioe = n |
-      last = n and completion = basicBooleanCompletion(false)
+      last.asExpr() = n and completion = basicBooleanCompletion(false)
       or
       last(ioe.getPattern(), last, NormalCompletion()) and completion = basicBooleanCompletion(true)
     )
     or
     // The last node of a node executed in post-order is the node itself.
-    n.(PostOrderNode).mayCompleteNormally() and last = n and completion = NormalCompletion()
+    // n.(PostOrderNode).mayCompleteNormally() and last = n and completion = NormalCompletion()
```
Comment (aschackmull) [inline java/ql/lib/semmle/code/java/ControlFlowGraph.qll:934]:
```suggestion
```

File: java/ql/lib/semmle/code/java/ControlFlowGraph.qll [hunk 5]
```diff
@@ -1117,17 +1194,19 @@ private module ControlFlowGraphImpl {
    * execution finishes with the given completion.
    */
   cached
-  ControlFlowNode succ(ControlFlowNode n, Completion completion) {
-    // Callables serve as their own exit nodes.
-    exists(Callable c | last(c.getBody(), n, completion) | result = c)
+  Node succ(Node n, Completion completion) {
+    // The successor of a callable is its exit node.
```
Comment (aschackmull) [inline java/ql/lib/semmle/code/java/ControlFlowGraph.qll:1198]:
```suggestion
    // After executing the callable body, the final node is the exit node.
```

File: java/ql/lib/semmle/code/java/ControlFlowGraph.qll [hunk 6]
```diff
@@ -1506,18 +1589,18 @@ private module ControlFlowGraphImpl {
...
 
   /** Gets a true-successor of the `ConditionNode`. */
-  ControlFlowNode getATrueSuccessor() { result = this.getABranchSuccessor(true) }
+  ControlFlow::Node getATrueSuccessor() { result = this.getABranchSuccessor(true) }
 
   /** Gets a false-successor of the `ConditionNode`. */
-  ControlFlowNode getAFalseSuccessor() { result = this.getABranchSuccessor(false) }
+  ControlFlow::Node getAFalseSuccessor() { result = this.getABranchSuccessor(false) }
 
   /** Gets the condition of this `ConditionNode`. This is equal to the node itself. */
```
Comment (aschackmull) [inline java/ql/lib/semmle/code/java/ControlFlowGraph.qll:1604]:
```suggestion
  /** Gets the condition of this `ConditionNode`. */
```

File: java/ql/lib/semmle/code/java/ControlFlowGraph.qll [hunk 7]
```diff
@@ -84,45 +84,107 @@
...
+
+  private class AstNode extends ExprParent {
+    AstNode() { this instanceof Expr or this instanceof Stmt }
+
+    Stmt getEnclosingStmt() {
+      result = this or
+      result = this.(Expr).getEnclosingStmt()
+    }
+
+    Node getCFGNode() { result.asExpr() = this or result.asStmt() = this }
```
[the remark this answers is not in the document]
Comment (aschackmull) [reply java/ql/lib/semmle/code/java/ControlFlowGraph.qll:200]: We should fix this.
```suggestion
    Node getCfgNode() { result.asExpr() = this or result.asStmt() = this }
```

File: java/ql/lib/semmle/code/java/ControlFlowGraph.qll [hunk 8]
```diff
@@ -1321,7 +1404,9 @@ private module ControlFlowGraphImpl {
         switchExpr = switch.(SwitchStmt).getExpr() or switchExpr = switch.(SwitchExpr).getExpr()
       |
         // From the entry point control is transferred first to the expression...
-        n = switch and result = first(switchExpr) and completion = NormalCompletion()
+        (n.asStmt() = switch or n.asExpr() = switch) and
```
Comment (aschackmull) [inline java/ql/lib/semmle/code/java/ControlFlowGraph.qll:1407]: Maybe just use `getAstNode`:
```suggestion
        n.getAstNode() = switch and
```

File: java/ql/lib/semmle/code/java/ControlFlowGraph.qll [hunk 9]
```diff
@@ -84,45 +84,122 @@ private import Completion
...
 
-  /** Gets a successor of this node that is neither an exception successor nor a jump (break, continue, return). */
-  ControlFlowNode getANormalSuccessor() {
-    result = succ(this, BooleanCompletion(_, _)) or
-    result = succ(this, NormalCompletion())
+    /**
+     * Get the most appropriate AST node for this control flow node, if any.
+     *
+     * This is needed for the equivalence relation on basic blocks in range
+     * analysis.
```
Comment (aschackmull) [inline java/ql/lib/semmle/code/java/ControlFlowGraph.qll:170]: That's not its only use. Also, it seems reasonable to have predicate like this without such a specific-sounding constraint.
```suggestion
     *
     * This is needed for the equivalence relation on basic blocks in range
     * analysis.
```

Comment (owen-mc) [reply java/ql/lib/semmle/code/java/ControlFlowGraph.qll:170]: Did you mean to suggest a different wording?

Comment (aschackmull) [reply java/ql/lib/semmle/code/java/ControlFlowGraph.qll:170]: Ah no, I meant to simply delete it. Edited.

Comment (owen-mc) [reply java/ql/lib/semmle/code/java/ControlFlowGraph.qll:170]: Would you expect every control flow node to be able to supply this? Or is it optional?

**Comment by owen-mc** [comment]
@aschackmull This is ready to re-review. I have accepted all your review suggestions. The failing test is because we need an internal PR as well.

File: java/ql/lib/semmle/code/java/ControlFlowGraph.qll (hunk 9 above)
Comment (aschackmull) [reply java/ql/lib/semmle/code/java/ControlFlowGraph.qll]: Hmm, there are two different potentially relevant predicates, one that provides the inverse of the `getControlFlowNode` on `Expr` and `Stmt`, and one that projects every Cfg node to an associated Ast node. I think I'd name the latter something with `project`, and let this predicate take the semantics of the former. Hence, I'd say that this is optional and that we should postfix its qldoc with the customary "if any".

**Review by aschackmull — APPROVED** [review]

**Review by aschackmull — APPROVED** [review]

**Review by aschackmull — DISMISSED** [review]

**Review by egregius313 — APPROVED** [review]
Re-approving

**Pull request merged**